packages/lit-html/src/directives/async-append.ts

```diff
@@ -0,0 +1,124 @@
+
+class AsyncAppendDirective extends AsyncDirective {
+  value?: AsyncIterable<unknown>;
+  reconnectResolver?: () => void;
+  reconnectPromise?: Promise<void>;
+
+  constructor(partInfo: PartInfo) {
+    super(partInfo);
+    if (partInfo.type !== PartType.CHILD) {
+      throw new Error('asyncAppend can only be used in child expressions');
+    }
+  }
+
+  render<T>(_value: AsyncIterable<T>, _mapper?: Mapper<T>) {
+    return noChange;
+  }
+
+  update(part: ChildPart, [value, mapper]: DirectiveParameters<this>) {
+    // If we've already set up this particular iterable, we don't need
+    // to do anything.
+    if (value === this.value) {
+      return;
+    }
+    this.value = value;
+    this.__iterate(part, mapper);
+    return noChange;
+  }
+
+  // Separate function to avoid an iffe in update; update can't be async
+  // because its return value must be `noChange`
+  private async __iterate(part: ChildPart, mapper?: Mapper<unknown>) {
+    let i = 0;
+    const {value} = this;
+    for await (let v of value!) {
+      // Check to make sure that value is the still the current value of
+      // the part, and if not bail because a new value owns this part
+      if (this.value !== value) {
+        break;
+      }
+
+      // If we were disconnected, pause until reconnected
+      if (this.reconnectPromise) {
+        await this.reconnectPromise;
+      }
+
+      // When we get the first value, clear the part. This lets the
+      // previous value display until we can replace it.
+      if (i === 0) {
+        clearPart(part);
+      }
+      // As a convenience, because functional-programming-style
+      // transforms of iterables and async iterables requires a library,
+      // we accept a mapper function. This is especially convenient for
+      // rendering a template for each item.
+      if (mapper !== undefined) {
+        v = mapper(v, i);
+      }
+      const newPart = insertPart(part);
+      setChildPartValue(newPart, v);
+      i++;
+    }
+  }
+
+  // Pause iteration while disconnected
+  disconnected() {
+    this.reconnectPromise = new Promise(
+      (resolve) => (this.reconnectResolver = resolve)
+    );
+  }
+
+  // Resume iteration when reconnected
+  reconnected() {
+    this.reconnectPromise = undefined;
+    this.reconnectResolver!();
+  }
+}
+
+/**
+ * A directive that renders the items of an async iterable[1], appending new
+ * values after previous values, similar to the built-in support for iterables.
+ *
+ * Async iterables are objects with a [Symbol.asyncIterator] method, which
+ * returns an iterator who's `next()` method returns a Promise. When a new
+ * value is available, the Promise resolves and the value is appended to the
+ * Part controlled by the directive. If another value other than this
+ * directive has been set on the Part, the iterable will no longer be listened
+ * to and new values won't be written to the Part.
+ *
+ * [1]: https://github.com/tc39/proposal-async-iteration
```

Comment:
Maybe like to this now: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of

Can't really find a better page though

Reply:
Done